### ISSUE TITLE
Haproxy uses a random backend when the actual backend for the route is not available

### ISSUE THREAD

**Issue by Sorunome** [haproxy/haproxy#717]
## Output of `haproxy -vv` and `uname -a`

```
root@sorunome ~ # haproxy -vv
HA-Proxy version 2.0.15-1~bpo10+1 2020/06/14 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-K58yyY/haproxy-2.0.15=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace

root@sorunome ~ # uname -a
Linux sorunome 4.19.0-8-amd64 #1 SMP Debian 4.19.98-1 (2020-01-26) x86_64 GNU/Linux
root@sorunome ~ # 
```

## What's the configuration?

```
global
        log /dev/log    local0
        log /dev/log    local1 notice
        chroot /var/lib/haproxy
        stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
        stats timeout 30s
        user haproxy
        group haproxy
        daemon
 
        tune.comp.maxlevel 6
        tune.bufsize 32768
        tune.chksize 32768
        tune.h2.header-table-size 8192
        tune.ssl.default-dh-param 2048
 
        # Default SSL material locations
        ca-base /etc/ssl/certs
        crt-base /etc/ssl/private
 
        # Default ciphers to use on SSL-enabled listening sockets.
        # For more information, see ciphers(1SSL). This list is from:
        #  https://hynek.me/articles/hardening-your-web-servers-ssl-ciphers/
        # An alternative list with additional directives can be obtained from
        #  https://mozilla.github.io/server-side-tls/ssl-config-generator/?server=haproxy
        ssl-default-bind-ciphers ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:RSA+AESGCM:RSA+AES:!aNULL:!MD5:!DSS
        ssl-default-bind-options no-sslv3
 
defaults
        log     global
        mode    http
        option  httplog
        option  dontlognull
        #option  http-server-close
        option http-keep-alive
        #option  forwardfor
        timeout connect 30000
#        timeout client  300000
        timeout client  60s
        timeout server  300000
        errorfile 400 /etc/haproxy/errors/400.http
        errorfile 403 /etc/haproxy/errors/403.http
        errorfile 408 /etc/haproxy/errors/408.http
        errorfile 500 /etc/haproxy/errors/500.http
        errorfile 502 /etc/haproxy/errors/502.http
        errorfile 503 /etc/haproxy/errors/503.http
        errorfile 504 /etc/haproxy/errors/504.http
 
backend ponies
    option forwardfor
    http-request set-header Host ponies.im
    server reverseport 127.0.0.1:1080
 
backend matrix
    redirect scheme https if !{ ssl_fc }
    option forwardfor
    server matrix 127.0.0.1:8008
 
backend nginx_localhost
    option forwardfor
    server nginx 127.0.0.1:2080
backend nginx_localhost_ssl
    option forwardfor
    server nginx 127.0.0.1:2443 ssl verify none
 
frontend soru
    # Options
    mode http
    compression algo gzip raw-deflate
    compression offload
 
    # Bind Rules
    bind 0.0.0.0:80
    bind :::80
    bind 0.0.0.0:443 ssl crt /etc/haproxy/certs alpn h2,http/1.1
    bind :::443 ssl crt /etc/haproxy/certs alpn h2,http/1.1
 
    acl sorunomede hdr_beg(host) -i sorunome.de
    acl sorunomedev6 hdr_beg(host) -i [2a01:4f8:120:4091::3]
    acl poniesim hdr_beg(host) -i ponies.im
    acl poniesimv6 hdr_beg(host) -i [2a01:4f8:120:4091::14]
 
    acl matrix path_beg /_matrix

    use_backend matrix if matrix sorunomede || matrix sorunomedev6
 
    use_backend ponies if poniesim || poniesimv6
 
    use_backend nginx_localhost_ssl if { ssl_fc }
    default_backend nginx_localhost
```

## Steps to reproduce the behavior

When the `matrix` backend is offline, the request is seemingly randomly routed to one of the other backends.
On the server:
```
root@sorunome ~ # curl http://127.0.0.1:8008
curl: (7) Failed to connect to 127.0.0.1 port 8008: Connection refused
```
Indicates that the backend is *actually* offline. As such, a 503 reply is expected. However, curl indicates that this isn't the case:

```
# Incorrect reply handled by nginx with ssl
sorunome@sorunome-desktop ~ $ curl https://sorunome.de/_matrix/client/versions -4
<html><body>You are being <a href="https://sorunome.de/_matrix/client/versions">redirected</a>.</body></html>
# Incorrect reply handled by nginx without ssl
sorunome@sorunome-desktop ~ $ curl https://sorunome.de/_matrix/client/versions -4
<html>
<head><title>302 Found</title></head>
<body bgcolor="white">
<center><h1>302 Found</h1></center>
<hr><center>nginx/1.14.2</center>
</body>
</html>
# Incorrect reply handled by nginx again with ssl
sorunome@sorunome-desktop ~ $ curl https://sorunome.de/_matrix/client/versions -4
<html><body>You are being <a href="https://sorunome.de/_matrix/client/versions">redirected</a>.</body></html>
# Incorrect reply handled by ponies backend
sorunome@sorunome-desktop ~ $ curl https://sorunome.de/_matrix/client/versions -4
{
    "unstable_features": {
        "m.id_access_token": true,
        "m.require_identity_server": false,
        "m.separate_add_and_bind": true,
        "org.matrix.e2e_cross_signing": true,
        "org.matrix.label_based_filtering": true,
        "org.matrix.msc2432": true
    },
    "versions": [
        "r0.0.1",
        "r0.1.0",
        "r0.2.0",
        "r0.3.0",
        "r0.4.0",
        "r0.5.0"
    ]
}
```

In the haproxy logs, the "backend / server" part is `matrix/matrix` for all of these replies, indicating they *should* route to the correct backend, while they actually don't.

There is only a single haproxy instance running.

## Actual behavior

See above

## Expected behavior

Haproxy should return 503

## Do you have any idea what may have caused this?

No idea

## Do you have an idea how to solve the issue?

The installed haproxy version is from "buster backports". Downgrading haproxy to the version from "buster" seems to work for now, though that is not a proper fix. This is the version output for the working haproxy from buster:

```
root@sorunome ~ # haproxy -vv
HA-Proxy version 1.8.19-1+deb10u2 2020/04/01
Copyright 2000-2019 Willy Tarreau <willy@haproxy.org>

Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-ap5Gic/haproxy-1.8.19=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-format-truncation -Wno-null-dereference -Wno-unused-label
  OPTIONS = USE_GETADDRINFO=1 USE_ZLIB=1 USE_REGPARM=1 USE_OPENSSL=1 USE_LUA=1 USE_SYSTEMD=1 USE_PCRE2=1 USE_PCRE2_JIT=1 USE_NS=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Encrypted password support via crypt(3): yes
Built with multi-threading support.
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with network namespace support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
```

**Comment by davidmehren** [comment]
This is also happening on Ubuntu 18.04 with `HA-Proxy version 2.0.15-1ppa1~bionic 2020/06/13`, the latest version from `ppa:vbernat/haproxy-2.0`.

**Comment by TimWolla** [comment]
@davidmehren To clarify: Are you independently seeing the issue or are you related to @Sorunome?

**Comment by Sorunome** [comment]
They are not related to me in any way

Am 27. Juni 2020 18:44:59 MESZ schrieb "Tim Düsterhus" <notifications@github.com>:

**Comment by michaelkaye** [comment]
I also experienced this issue; requests logged as being routed to one backend/server but actually being routed to an entirely different backend/server.

For us it was triggered for us with the upgrade from `2.0.14-1~bpo10+1` to `2.0.15-1~bpo10+1`, from debian backports - rolling back to `2.0.14-1~bpo10+1` with identical configuration fixed the issue.

I've been trying to replicate it outside of production for the last few days, with no success, but I didn't specifically try it with backends that were unavailable - we were just seeing the behaviour start at random during the normal flow of traffic. Some of our backends do stop responding from time to time, so it's not unreasonable that this is triggered by a backend that isn't responding to connections.

Tcpdumps of the traffic from HAProxy to the backend servers also didn't appear odd - We use keepalives and so have a good number of http requests down each tcp connection.  I didn't at any point see http requests that should have been routed to two different backends ending up sharing a single TCP connection. Each TCP connection was handling just one backend's traffic - it's just like some of those tcp connections had been connected to the wrong backend server.

(This is independent again of either of the above reports)

**Comment by wtarreau** [comment]
I couldn't reproduce any of this and even after reading the code in details, I see that we'll take either *THE* matching backend, or stay into the frontend, which will indeed produce a 503. So for me it looks extremely strange, and a reliable reproducer would be much appreciated.

@Sorunome if you can reproduce it on an isolated node, I'd like to see a dump of the stats before and after the failing request (just issue a "show stats typed" on the CLI). We might find an undesired redispatch or any such unexpected thing that might help understand what's happening.

**Comment by lukastribus** [comment]
A full repro including bisect has been posted on the ML:

https://www.mail-archive.com/haproxy@formilux.org/msg37850.html

commit 7b69c91e7d9ac6d7513002ecd3b06c1ac3cb8297 is only in 2.0 (it's not a backport, but a specific patch for 2.0).

**Comment by capflam** [comment]
I pushed a fix for the 2.0. See [e8d24235](http://git.haproxy.org/?p=haproxy-2.0.git;a=commit;h=e8d2423543314a586f7c76087cb1335e6a6dcef6) and [d982a8e8](http://git.haproxy.org/?p=haproxy-2.0.git;a=commit;h=d982a8e802ef0cea2a903a60f355d8513d7dcd53).
